bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/values/ReadOnlyUtils.java

```diff
@@ -124,6 +126,10 @@ private static Type getAvailableImmutableType(Type type) {
             return ((BIntersectionType) type).getEffectiveType();
         }
 
+        if (type.getTag() == TypeTags.TYPE_REFERENCED_TYPE_TAG) {
+            return getAvailableImmutableType(((ReferenceType) type).getReferredType());
```

Comment:
Should this be call before checking the intersection type?

Reply:
Since this is a recursive call to the same method, it will not affect the `IntersectionType` related logic.